python/test/unit/mesh/test_mesh_utils.py

```diff
@@ -190,18 +328,47 @@ def test_topology_partition():
     # NOTE: Maybe we can ensure that the 'global cells' are in the same
     # order as the owned cells (maybe they are already) to avoid the
     # need for global_index_nodes
-    cell_nodes, global_index_nodes = cpp.mesh.exchange(cpp.MPI.comm_world,
-                                                       cells1, dest, set(src))
+    #
+    # NOTE: This could be optimised as we have earlier computed which
+    # processes own the cells this process needs.
+    cell_nodes, global_index_cell = cpp.mesh.exchange(cpp.MPI.comm_world,
+                                                      cells_global, dest, set(src))
     assert cell_nodes.num_nodes == cells.num_nodes
-    assert global_index_nodes == original_cell_index
+    assert global_index_cell == original_cell_index
 
-    # Check that number of dofs is equal to number of 'nodes' in the input
-    assert dofmap.shape == cell_nodes.array().shape
+    # Check that number of dofs is equal to number of geometry 'nodes'
+    # from the mesh data input
+    assert dofmap.array().shape == cell_nodes.array().shape
 
-    # Build list of unique node indices
+    # Build list of unique (global) node indices from adjacency list
+    # (geometry nodes)
     indices = np.unique(cell_nodes.array())
 
-    # Fetch node coordinates
+    # Fetch node coordinates by global index from other ranks. Order of
+    # coords matches order of the indices in 'indices'
     coords = cpp.mesh.fetch_data(cpp.MPI.comm_world, indices, x)
-    for index, value in zip(indices, coords):
-        print("Index, x:", index, value)
+
+    # Compute local-to-global map from local indices in dofmap to the
+    # corresponding global indices in cell_nodes
+    l2g = cpp.mesh.compute_local_to_global_links(cell_nodes, dofmap)
+
+    # Compute local (dof) to local (position in coords) map from (i)
+    # local-to-global for dofs and (ii) local-to-global for entries in
+    # coords
+    l2l = cpp.mesh.compute_local_to_local(l2g, indices)
+
+    # Build coordinate dof array
+    x_g = coords[l2l]
+
+    # Create Geometry
+    geometry = cpp.mesh.Geometry(dof_index_map, dofmap, x_g, l2g, order)
+
+    # Create mesh
+    mesh = cpp.mesh.Mesh(cpp.MPI.comm_world, topology, geometry)
+
+    # Write mesh to file
```

Comment:
This was for debugging I guess.

Reply:
Also picks up some issues sometimes, but mainly for visual inspection.